### ISSUE TITLE
static variables incorrectly read

### ISSUE THREAD

**Issue by tjcorona** [gccxml/pygccxml#64]
in pygccxml/pygccxml/parser/scanner.py: 559, should 

```
type_qualifiers.has_static = attrs.get(XML_AN_EXTERN, False)
```

be

```
type_qualifiers.has_static = attrs.get(XML_AN_STATIC, False)
```

?

**Comment by iMichka** [comment]
Hi

Thanks for reporting this. This line has been like this since 2006, so if it is a bug, it's an old one. But it may also be that the previous xml general (gccxml) returned some weird output, and this was a way to "make it work".

Making the proposed change sounds sane, but it makes one of the tests fail:
https://github.com/gccxml/pygccxml/blob/develop/unittests/declarations_tester.py#L59-L65

Interestingly the tested variable has a comment from the initial author of pygccxml:
https://github.com/gccxml/pygccxml/blob/develop/unittests/data/declarations_variables.hpp#L14-L15
This sounds like there was already something fishy here 10 years ago.

I will need to have a close look at the xml output, but it may take some days as I have other things to finish first. I hope I can have a look this weekend.

**Comment by iMichka** [comment]
Here is some doc for myself:
http://faq.cprogramming.com/cgi-bin/smartfaq.cgi?answer=1321396234&id=1043284376

I had a look and the relevant lines in CastXML are here for variables:
https://github.com/CastXML/CastXML/blob/master/src/Output.cxx#L1878-L1883
And here for functions:
https://github.com/CastXML/CastXML/blob/master/src/Output.cxx#L1501-L1503
https://github.com/CastXML/CastXML/blob/master/src/Output.cxx#L1519-L1521

I am continuing my investigations. For the moment it looks like that even if you use the static qualifier, clang will detect as extern.

Do you have a specific c++ code example that is not working, so I could write a small test for it and add it to the test suite?

**Comment by dermont123** [comment]
Hi,
I'm running into a problem with this as well, the values of static being returned are different between gccxml and castxml. Latest verion 1.8.4 downloaded a few days ago.

static const Vector ZERO;
static int i =1;
----------------------
g++-4.9 gccxml
----------------------
XML_AN_NAME   =  ZERO
XML_AN_EXTERN =  1
XML_AN_STATIC =  1
XML_AN_MUTABLE =  False

has_static = True
has_extern = True
----------------------
g++6.2 castxml
----------------------
XML_AN_NAME   =  ZERO
XML_AN_EXTERN =  False
XML_AN_STATIC =  1
XML_AN_MUTABLE =  False
has_static = False
has_extern = False

Castxml returns a list of NOT static vars find_noncopyable_vars (statics) , py++ checks if noncopyable or find_noncopyable_vars not empty and defines class noncopyable '::boost::noncopyable'. In addition since the vars are not static there is no way to check/exclude and leads to undefined symbols when linking. Py++ also appears to flip the static value based on the extern value (outwith how it is read in the pygccxml scanner).

I've updated and attached a couple of your unittest. 

I had to make minor updates to py++ (to run pygccxml1.8.4) / gccxml (for g++4.9), so maybe it's something change I made.

Thanks for the work in updating pygccxml for castxml.

[probs.zip](https://github.com/gccxml/pygccxml/files/776982/probs.zip)

**Comment by iMichka** [comment]
Thanks for the details. I will have a look at this during the weekend.

For pyplusplus, make sure you are using the version here: https://bitbucket.org/ompl/pyplusplus/
It's the most up to date project. They will be pleased to get pull request if you have any fixes that could help them.

**Comment by iMichka** [comment]
Okay I think I have found out what the problem was. In former times (with `GCC-XML`), the `static` vs `extern` detection was buggy. Now, using `CastXML`, the output seems to look better.

I made two changes.

- First one is to make sure the type_qualifiers class knows the difference between `static` and `extern`. I left the behaviour as-is for `GCC-XML`, which will be dropped soon anyway.
See: https://github.com/gccxml/pygccxml/commit/e697d33c3310a2ccb7ff500a51a17ae49f45dae9

- I double checked the `is_noncopyable` code, and added an example with a `static const` variable. In that case the class is copyable. See: https://github.com/gccxml/pygccxml/commit/d3780a9f53e682e437cef82acc2b2d362c5dceb9

You can test all this at once using the `1.8.5` branch:
https://github.com/gccxml/pygccxml/tree/hotfix/v1.8.5

Please tell me if this okay for you, and I'll make a new release with these changes. Thanks.

**Comment by dermont123** [comment]
Thanks will try it out in the next day or so.

**Comment by dermont123** [comment]
Tests OK here, thanks. One slight problem it failed on a pointer_t  with no attribute is_abstract. I just added the try/except as in the previous version:
`
diff --git a/pygccxml/declarations/type_traits_classes.py b/pygccxml/declarations/type_traits_classes.py
index 3b67284..dc1969f 100644
--- a/pygccxml/declarations/type_traits_classes.py
+++ b/pygccxml/declarations/type_traits_classes.py
@@ -672,9 +672,12 @@ def is_noncopyable(class_):
     if is_union(class_):
         return False
 
-    if class_decl.is_abstract:
-        logger.debug(true_header + "abstract client")
-        return True
+    try:
+        if class_decl.is_abstract:
+            logger.debug(true_header + "abstract client")
+            return True
+    except:
+        return False
 
     # if class has public, user defined copy constructor, than this class is
     # copyable
`

**Comment by iMichka** [comment]
I made change for that case, which was not covered by the tests: https://github.com/gccxml/pygccxml/commit/8653705798bea652bd2f33fe4248468340c50165

I have some ideas for performance improvements in these methods, but that is a different topic and surely not to be done in this release.

Could you please again the 1.8.5 branch?

**Comment by dermont123** [comment]
Thanks I built my project without problems. 

I also ran with the headers only, including everything, py++ fails when checking the array_item_type for  unknown_t \ enumeration_t in type_traits_classes::is_noncopyable and checking the pointer type for a free_function_type_t.

`
AttributeError: 'enumeration_t' object has no attribute 'is_abstract'
AttributeError: 'free_function_type_t' object has no attribute 'is_abstract'
...
`
Also passing a ::std::basic_ostream type to find_noncopyable_vars, results in an endless loop. Sorry I'll try to create a test case for this.
`
find_noncopyable_vars()
    if class_traits.is_my_case(type_)
        -> is_noncopyable()
            -> __is_noncopyable_single()
                find_noncopyable_vars()
`

[traits.zip](https://github.com/gccxml/pygccxml/files/794660/traits.zip)

**Comment by iMichka** [comment]
These seem to be different errors.

I just released version 1.8.5 with the changes for the static/extern qualifiers; and the pointer_t problem. Because there was another fix that needed to be addressed and the release couldn't wait.

Please open new separate bugs for the two other problems. There we can discuss them and I will try to address them. Thanks for all the testing, it is really appreciated.
